src/lib.rs

```diff
@@ -0,0 +1,534 @@
+    checking_interval: Duration,
+}
+
+impl Memtester {
+    // TODO: Memtester without given base_ptr, ie. take care of memory allocation as well
+
+    /// Create a Memtester containing all test types in random order
+    pub fn all_tests_random_order(args: MemtesterArgs) -> Memtester {
+        let mut test_types = vec![
+            MemtestType::TestOwnAddress,
+            MemtestType::TestRandomVal,
+            MemtestType::TestXor,
+            MemtestType::TestSub,
+            MemtestType::TestMul,
+            MemtestType::TestDiv,
+            MemtestType::TestOr,
+            MemtestType::TestAnd,
+            MemtestType::TestSeqInc,
+            MemtestType::TestSolidBits,
+            MemtestType::TestCheckerboard,
+            MemtestType::TestBlockSeq,
+        ];
+        test_types.shuffle(&mut thread_rng());
+
+        Self::from_test_types(args, test_types)
+    }
+
+    /// Create a Memtester with specified test types
+    pub fn from_test_types(args: MemtesterArgs, test_types: Vec<MemtestType>) -> Memtester {
+        Memtester {
+            test_types,
+            timeout: args.timeout,
+            require_memlock: args.require_memlock,
+            allow_working_set_resize: args.allow_working_set_resize,
+            allow_mem_resize: args.allow_mem_resize,
+            allow_multithread: args.allow_multithread,
+            allow_early_termination: args.allow_early_termination,
+        }
+    }
+
+    /// Run the tests, possibly after locking the memory
+    pub fn run(&self, memory: &mut [usize]) -> anyhow::Result<MemtestReportList> {
+        // TODO: Should have a minimum memory length so that we don't UB when `memory.len()` is too small
+        let mut timeout_checker = TimeoutChecker::new(Instant::now(), self.timeout);
+
+        // TODO: the linux memtester aligns base_ptr before mlock to avoid locking an extra page
+        //       By default mlock rounds base_ptr down to nearest page boundary
+        //       Not sure which is desirable
+
+        if self.require_memlock {
+            #[cfg(windows)]
+            let working_set_sizes = if self.allow_working_set_resize {
+                Some(
+                    windows::replace_set_size(size_of_val(memory))
+                        .context("Failed to replace process working set size")?,
+                )
+            } else {
+                None
+            };
+
+            let (memory, mlocked) = match memory_resize_and_lock(memory, self.allow_mem_resize) {
+                Ok(resized_memory) => (resized_memory, true),
+                Err(e) => {
+                    // TODO: Returning without restoring set size?
+                    bail!(e.context(
+                        "Failed memory locking when it is required, abort running Memtester",
+                    ));
+                }
+            };
+
+            let reports = self.run_tests(memory, &mut timeout_checker)?;
+
+            if let Err(e) = memory_unlock(memory) {
```

Comment:
I will leave it to you to implement `Drop` here as well (to make sure you don't forget to unlock the memory!)